### PR TITLE
AVL tree library

### PR DESCRIPTION
**Pull request by mllwchrry**
- [ ] Since this PR suggests a **bug fix**, the tests have been added and the coverage is 100%.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

### CODE REVIEW

**Review by Hrom131 — APPROVED** [review]
LGTM

File: contracts/mock/libs/data-structures/AvlTreeMock.sol [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        if (value2_.length < minLength) minLength = value2_.length;
+
+        for (uint i = 0; i < minLength; i++) {
+            if (value1_[i] < value2_[i]) return -1;
+            else if (value1_[i] > value2_[i]) return 1;
+        }
+
+        if (value1_.length < value2_.length) return -1;
+        else if (value1_.length > value2_.length) return 1;
+        else return 0;
```
Comment (Arvolear) [inline contracts/mock/libs/data-structures/AvlTreeMock.sol:256]: Style please

File: contracts/libs/data-structures/AvlTree.sol [hunk 2]
```diff
@@ -0,0 +1,823 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {TypeCaster} from "../utils/TypeCaster.sol";
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:5]:
```suggestion
import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";

import {TypeCaster} from "../utils/TypeCaster.sol";
```

File: contracts/libs/data-structures/AvlTree.sol [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+
+/**
+ * @notice AVL Tree module
+ *
+ * This library provides implementation of three sets with dynamic key types:
+ * `UintAVL`, `Bytes32AVL` and `Bytes32AVL`.
+ *
+ * Each element in the tree contains a bytes `value` field to allow storing different types
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:13]: Not sure if we need bytes

File: contracts/libs/data-structures/AvlTree.sol [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+
+    function _defaultComparator(
+        bytes32 key1_,
+        bytes32 key2_,
+        bytes memory,
+        bytes memory
+    ) private pure returns (int8) {
+        if (key1_ < key2_) return -1;
+        if (key1_ > key2_) return 1;
+        return 0;
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:809]: Style

File: contracts/libs/data-structures/AvlTree.sol [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+        returns (function(bytes32, bytes32, bytes memory, bytes memory) view returns (int8))
+    {
+        return tree.isCustomComparatorSet ? tree.comparator : _defaultComparator;
+    }
+
+    function _defaultComparator(
+        bytes32 key1_,
+        bytes32 key2_,
+        bytes memory,
+        bytes memory
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:805]: Comparator for values? Interesting...

File: contracts/libs/data-structures/AvlTree.sol [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+     *     Internal Tree     *
+     *************************
+     */
+
+    struct Node {
+        bytes32 key;
+        bytes value;
+        uint256 height;
+        bytes32 left;
+        bytes32 right;
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:517]:
```suggestion
        bytes32 key;
        bytes32 value;
        uint64 height;
        uint64 parent;
        uint64 left;
        uint64 right;
```

And store the nodes with incremental indexes. Yep, the `search` function will be logarithmic but we 

1. Save storage and gas
2. Able to implement `iterator` design pattern to traverse the tree

File: contracts/libs/data-structures/AvlTree.sol [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+    }
+
+    struct Tree {
+        bytes32 root;
+        uint256 treeSize;
+        bool isCustomComparatorSet;
+        mapping(bytes32 => Node) tree;
+        function(bytes32, bytes32, bytes memory, bytes memory) view returns (int8) comparator;
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:525]:
```suggestion
        uint64 root;
        uint64 treeSize;
        bool isCustomComparatorSet;
        mapping(bytes32 => Node) tree;
        function(bytes32, bytes32, bytes memory, bytes memory) view returns (int8) comparator;
```

File: contracts/libs/data-structures/AvlTree.sol [hunk 8]
```diff
@@ -0,0 +1,823 @@
...
+        bytes32 node_
+    ) private view returns (bytes32) {
+        while (_tree[node_].right != 0) {
+            node_ = _tree[node_].right;
+        }
+
+        return node_;
+    }
+
+    function _getTraversal(
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:717]: Using the `iterator` pattern will delegate the traversal responsibility to the special object (struct in our case). We will then implement a "traversal" library for these objects.

File: contracts/libs/data-structures/AvlTree.sol [hunk 9]
```diff
@@ -0,0 +1,823 @@
...
+        bytes32 key2_,
+        bytes memory,
+        bytes memory
+    ) private pure returns (int8) {
+        if (key1_ < key2_) return -1;
+        if (key1_ > key2_) return 1;
+        return 0;
+    }
+
+    function _asAddress(bytes32 from_) private pure returns (address to_) {
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:812]: Let's use high-level solidity casting. This adds an additional jump instruction

File: contracts/libs/data-structures/AvlTree.sol [hunk 10]
```diff
@@ -0,0 +1,823 @@
...
+    }
+
+    /**
+     * @notice The function to return the key of the root element of the address tree.
+     * Complexity is O(1).
+     *
+     * @param tree self.
+     * @return The key of the root element of the uint256 tree.
+     */
+    function root(AddressAVL storage tree) internal view returns (address) {
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:457]: What is the purpose of AVL root? I see no business benefits from this function

File: contracts/libs/data-structures/AvlTree.sol [hunk 11]
```diff
@@ -0,0 +1,823 @@
...
+     * @return The value associated with the key.
+     */
+    function getValue(
+        AddressAVL storage tree,
+        address key_
+    ) internal view returns (bytes storage) {
+        require(
+            _search(tree._tree, _asBytes32(key_)),
+            "AvlTree: node with such key doesn't exist"
+        );
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:424]: Oh nonono. Such reverts are extremely painful for the end users. 

Have two values returned. The actual value and the element's existence (to handle the empty case). The `search` function can be removed then. 

See openzeppelin's enumerable map

File: contracts/libs/data-structures/AvlTree.sol [hunk 12]
```diff
@@ -0,0 +1,823 @@
...
+    /**
+     * @notice The function to retrieve the maximum key in the address tree.
+     * Complexity is O(log(n)), where n is the number of elements in the tree.
+     *
+     * @param tree self.
+     * @return The maximum key in the tree.
+     */
+    function getMax(AddressAVL storage tree) internal view returns (address) {
+        return _asAddress(_getMax(tree._tree.tree, tree._tree.root));
+    }
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:448]: Ahh dunno... Probably we don't need them either.

**Review by Arvolear — CHANGES_REQUESTED** [review]
🦄

File: contracts/libs/data-structures/AvlTree.sol [hunk 13]
```diff
@@ -682,142 +707,91 @@ library AvlTree {
...
-        mapping(bytes32 => Node) storage _tree,
-        bytes32 node_,
-        bytes32[] memory keys_,
-        uint256 index_
-    ) private view returns (uint256) {
-        if (node_ == 0) {
-            return index_;
+        if (root_ == 0) {
+            return Traversal.Iterator({treeMappingSlot: treeMappingSlot_, currentNode: 0});
         }
```
Comment (Arvolear) [inline contracts/libs/data-structures/AvlTree.sol:762]: Do you need this `if`?

**Review by Arvolear — APPROVED** [review]
Bombarda

**Pull request merged**